### PR TITLE
XW-204 | Weibo widget

### PR DESCRIPTION
**Pull request by bognix**
/cc @vforge @rogatty

### CODE REVIEW

File: extensions/wikia/WeiboTag/TagBuilderHelper.class.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?php
+class TagBuilderHelper {
```
Comment (vforge) [inline extensions/wikia/WeiboTag/TagBuilderHelper.class.php]: Can you make sure it's re-useable by other tags?
or move it to WikiaParserTagController

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+	public static function onParserFirstCallInit( Parser $parser ) {
+		$parser->setHook( self::PARSER_TAG_NAME, [ new static(), 'renderTag' ] );
+		return true;
+	}
+
+	public function renderTag( $input, array $args, Parser $parser, PPFrame $frame ) {
+		$this->tagBuildSource = self::TAG_SRC . $this->tagBuilderHelper->buildTagSourceQueryParams( $this->tagSourceAllowedParamsWithDefaults, $args );
+		$iframe = Html::element( 'iframe',  $this->buildTagAttributes($args), wfMessage( 'weibotag-could-not-render' )->text() );
+		if ( $this->app->checkSkin( [ 'wikiamobile', 'mercury' ] ) ) {
```
Comment (vforge) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php]: please create isMobile function TagBuilderHelper instead

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			return Html::rawElement( 'script',  ['type' => 'x-wikia-widget'], $iframe );
```
Comment (vforge) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php]: What's the reason of wrapping data into another element?

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+		} else {
+			return $iframe;
+		}
+	}
+
+	private function buildTagAttributes($args) {
```
Comment (vforge) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php]: This also can be abstracted

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	private $tagAllowedAttributes;
+	private $tagDefaultAttributes;
+	private $tagBuildSource;
+	private $tagBuilderHelper;
+
+
+	public function __construct() {
+		parent::__construct();
+
+		$this->tagSourceAllowedParamsWithDefaults = [
```
Comment (vforge) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php]: Please consider using `const` here.

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+			'language' => '',
+			'sense' => '',
+			'showinfo' => '1',
+			'showtitle' => '1',
+			'uids' => '',
+			'verified' => '',
+			'wide' => '',
+		];
+
+		$this->tagAllowedAttributes = [
```
Comment (vforge) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php]: Please consider using `const` here.

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+			'width',
+			'height',
+			'scrolling',
+			'frameborder',
+			'style'
+		];
+
+		$this->tagDefaultAttributes = [
```
Comment (vforge) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php]: Please consider using `const` here.

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+			'verified' => '',
+			'wide' => '',
+		];
+
+		$this->tagAllowedAttributes = [
+			'width',
+			'height',
+			'scrolling',
+			'frameborder',
+			'style'
```
Comment (vforge) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php]: missing comma

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+		];
+
+		$this->tagDefaultAttributes = [
+			'data-wikia-widget' => self::PARSER_TAG_NAME,
+			'sandbox' => 'allow-scripts allow-same-origin',
+			'seamless' => 'seamless'
```
Comment (vforge) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php]: missing comma

File: extensions/wikia/WeiboTag/TagBuilderHelper.class.php [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+	public function buildTagAttributes(array $allowedAttrs, array $userAttrs) {
+		$attributes = [];
+
+		foreach ( $allowedAttrs as $attributeName ) {
+			if ( isset( $userAttrs[$attributeName] ) ) {
+				if ($attributeName === 'style') {
+					$attributes['style'] = Sanitizer::checkCss( $userAttrs['style'] );
+				} else {
+					$attributes[$attributeName] = $userAttrs[$attributeName];
```
Comment (vforge) [inline extensions/wikia/WeiboTag/TagBuilderHelper.class.php]: please consider using strip_tags

File: extensions/wikia/WeiboTag/TagBuilderHelper.class.php [hunk 11]
```diff
@@ -0,0 +1,28 @@
+<?php
+class TagBuilderHelper {
+
+	public function buildTagSourceQueryParams( array $allowedParams, array $userParams ) {
+		foreach ( array_keys( $allowedParams ) as $name ) {
+			if ( isset( $userParams[$name] ) ) {
+				$allowedParams[$name] = $userParams[$name];
+			}
+		}
+		return http_build_query( $allowedParams );
```
Comment (vforge) [inline extensions/wikia/WeiboTag/TagBuilderHelper.class.php]: please consider using strip_tags

**Comment by vforge** [comment]
LGTM

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 12]
```diff
@@ -0,0 +1,76 @@
+<?php
+class WeiboTagController extends WikiaParserTagController {
+	const PARSER_TAG_NAME = 'weibo';
+	const TAG_SRC = 'http://widget.weibo.com/relationship/bulkfollow.php?';
+	const TAG_SOURCE_ALLOWED_PARAMS_WITH_DEFAULTS = [
+		'color' => '',
```
Comment (rogatty) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php:6]: Weibo's default is `C2D9F2,FFFFFF,0082CB,666666`. Does setting it to an empty string overrides them? The same goes for all params. See http://jssdk.sinaapp.com/widget/bulkfollow.php#wb for partial list of defaults.

Comment (bognix) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php:6]: when you send empty weibo applies their defaults

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+		'count' => '',
+		'dpc' => '',
```
Comment (rogatty) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php]: We don't know what it does (seemingly nothing) so there is no way to document it. Should we remove it?

File: extensions/wikia/WeiboTag/WeiboTagController.class.php (hunk 12 above)
Comment (bognix) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php:6]: I changed WikiaTagBuilderHelper so it doesn't send empty params

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+			wfMessage( 'weibotag-could-not-render' )->text()
+		);
+
+		return $this->iFrameHelper->wrapForMobile( $iFrameHTML );
+	}
+
+	private function buildTagAttributes( $args ) {
+		$attributes = $this->tagBuilderHelper->buildTagAttributes( self::TAG_ALLOWED_ATTRIBUTES, $args );
+		$attributes['src'] = $this->tagBuildSource;
+		return array_merge( $attributes, self::TAG_DEFAULT_ATTRIBUTES );
```
Comment (hakubo) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php]: can you add on \n before return

I think it's easier to read

**Comment by hakubo** [comment]
LGTM

File: extensions/wikia/SoundCloudTag/SoundCloudTagController.class.php [hunk 15]
```diff
@@ -28,6 +28,7 @@ class SoundCloudTagController extends WikiaParserTagController {
 		'data-wikia-widget' => self::TAG_NAME,
 		'scrolling' => 'no',
 		'frameborder' => 'no',
+		'height' => '465',
```
Comment (rogatty) [inline extensions/wikia/SoundCloudTag/SoundCloudTagController.class.php:32]: Can you add a comment explaining this default?

File: extensions/wikia/WeiboTag/WeiboTagController.class.php [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+	];
+	const TAG_ALLOWED_ATTRIBUTES = [
+		'width',
+		'height',
+	];
+	const TAG_DEFAULT_ATTRIBUTES = [
+		'data-wikia-widget' => self::PARSER_TAG_NAME,
+		'scrolling' => 'no',
+		'frameborder' => '0',
+		'height' => '500',
```
Comment (rogatty) [inline extensions/wikia/WeiboTag/WeiboTagController.class.php:25]: Can you add a comment explaining this default?

File: extensions/wikia/WeiboTag/WeiboTag.i18n.php [hunk 17]
```diff
@@ -0,0 +1,12 @@
+<?php
+$messages = [];
+
+$messages['en'] = [
+	'weibotag-desc' => 'Weibo widget integration for the Wikia platform',
```
Comment (rogatty) [inline extensions/wikia/WeiboTag/WeiboTag.i18n.php]: `weibo-tag`?

File: extensions/wikia/SoundCloudTag/SoundCloudTag.i18n.php [hunk 18]
```diff
@@ -2,11 +2,11 @@
 $messages = [];
 
 $messages['en'] = [
-	'soundcloudtag-desc' => 'SoundCloud widget integration for the Wikia platform',
-	'soundcloudtag-could-not-render' => 'Failed to render the SoundCloud widget.',
+	'soundcloud-tag-desc' => 'SoundCloud widget integration for the Wikia platform',
+	'soundclouds-tag-could-not-render' => 'Failed to render the SoundCloud widget.',
```
Comment (rogatty) [inline extensions/wikia/SoundCloudTag/SoundCloudTag.i18n.php]: Typo, should be `soundcloud`

**Pull request merged**